content/docs/esc/environments/at-scale.md

```diff
@@ -0,0 +1,285 @@
+
+<img src="/images/docs/guides/esc/common-env-permissions.png"><br>
+
+### Environments in Applications
+
+An application engineer on the identity team, Jane, can now see the environment and use it in their `auth-bucket` project. This program is small, focused solely on creating an s3 bucket. Jane also takes advantage of the ESC SDK which allows the program to resolve and reference the AWS OIDC credentials at runtime to authenticate with AWS.
+
+{{% chooser language "go" / %}}
+
+{{% choosable language go %}}
+
+```go
+package main
+
+import (
+	"fmt"
+	"os"
+
+	esc "github.com/pulumi/esc-sdk/sdk/go"
+	"github.com/pulumi/pulumi-aws/sdk/v6/go/aws"
+	"github.com/pulumi/pulumi-aws/sdk/v6/go/aws/s3"
+	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
+)
+
+const (
+	projName = "common"
+	envName  = "aws"
+)
+
+func valueForEnvKey(envID string, envDef map[string]interface{}, key string, isSecret bool) (string, error) {
+	keyType := "configuration"
+	if isSecret {
+		keyType = "secret"
+	}
+	val, ok := envDef[key].(string)
+	if !ok {
+		return "", fmt.Errorf("%s key '%s' not found in environment %s", keyType, key, envID)
+	}
+	return val, nil
+}
+
+func getAwsProviderFromEnv(ctx *pulumi.Context, env map[string]any) (*aws.Provider, error) {
+	envID := fmt.Sprintf("%s/%s", projName, envName)
+	awsLogin, ok := env["login"]
+	if !ok {
+		return nil, fmt.Errorf("configuration key 'login' not found in environment %s", envID)
+	}
+
+	awsLoginMap, ok := awsLogin.(map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("configuration key 'login' in environment %s is not map[string]string", envID)
+	}
+
+	awsAccessKeyID, err := valueForEnvKey(envID, awsLoginMap, "accessKeyId", true)
+	if err != nil {
+		return nil, err
+	}
+
+	awsSecretAccessKey, err := valueForEnvKey(envID, awsLoginMap, "secretAccessKey", true)
+	if err != nil {
+		return nil, err
+	}
+
+	awsSessionTkn, err := valueForEnvKey(envID, awsLoginMap, "sessionToken", true)
+	if err != nil {
+		return nil, err
+	}
+
+	awsRegion, err := valueForEnvKey(envID, env, "region", false)
+	if err != nil {
+		return nil, err
+	}
+
+	return aws.NewProvider(ctx, "awsProvider", &aws.ProviderArgs{
+		AccessKey: pulumi.StringPtr(awsAccessKeyID),
+		SecretKey: pulumi.StringPtr(awsSecretAccessKey),
+		Region:    pulumi.StringPtr(awsRegion),
+		Token:     pulumi.StringPtr(awsSessionTkn),
+	})
+}
+
+func main() {
+	pulumi.Run(func(ctx *pulumi.Context) error {
+		accessToken := os.Getenv("PULUMI_ACCESS_TOKEN")
+		orgName := ctx.Organization()
+		configuration := esc.NewConfiguration()
+		escClient := esc.NewClient(configuration)
+		authCtx := esc.NewAuthContext(accessToken)
+
+        // Always reference the `stable` version of this environment
+		_, values, err := escClient.OpenAndReadEnvironmentAtVersion(authCtx, orgName, projName, envName, "stable")
+		if err != nil {
+			return fmt.Errorf("failed to open environment: %v", err)
+		}
+
+		awsProvider, err := getAwsProviderFromEnv(ctx, values)
```

Comment:
I think this example can be simplified because the default aws provider will pick up these env vars automatically if the environment is imported by the stack config.

I think using the ESC SDK within a stack makes the example appear more complicated, and we might be better served having a different non-stack example of using the SDK instead.

Reply:
👍 yeah I did have the thought this is a somewhat exaggerated example meant to show using JIT secrets in a practical application.

I ended up sticking with it because while the provider will use the environment variables by default you have to set those which when performing updates locally means you have to open the environment as an explicit step. With this you don't have to do any of that. Just as long as your PAT is set correctly.

Reply:
is an explicit open really required? just having the environment in the config seems to work for me 😕 
```
environment:
  - dev-sandbox
```

```
import * as aws from "@pulumi/aws";
const bucket = new aws.s3.BucketV2("my-bucket");
export const bucketName = bucket.id;
```